File: examples/recursive/mergesort.rb
require 'forkjoin'

class MergeSort < ForkJoin::Task
  def initialize(array, low, high)
    @array, @low, @high = array, low, high
  end

  def call
    size = @high - @low
    if size <= 8
      @array[@low, size] = @array[@low, size].sort
    else
      middle = @low + (size >> 1)
      (f = MergeSort.new(@array, middle, @high)).fork
      MergeSort.new(@array, @low, middle).call
      f.join
      merge(middle)
    end
  end

private

  def merge(middle)
    return if @array[middle - 1] < @array[middle]
    copy_size = @high - @low
    copy_middle = middle - @low
    copy = @array[@low, copy_size]
    p = 0
    q = copy_middle
    @low.upto(@high - 1) do |i|
      if q >= copy_size || (p < copy_middle && copy[p] < copy[q])
        @array[i] = copy[p]
        p += 1
      else
        @array[i] = copy[q]
        q += 1
      end
    end
  end
end

n = ARGV.shift.to_i

array = Array.new(n)
0.upto(n - 1) do |idx|
  array[idx] = rand(2**32) - 2**31
end

org = array.dup

start = Time.now.to_f
pool = ForkJoin::Pool.new
pool.submit(MergeSort.new(array, 0, array.size)).join
puts "%f [msec]" % ((Time.now.to_f - start) * 1000)

if array != org.sort
  # Huh? On my 4 core machine, it *sometimes* fail.
  # Runs fine on my 1 core machine.
  raise "Merge-sort failed!"
end

File: examples/recursive/MergeSort.java
import java.util.Arrays;
import java.util.Random;
import java.util.concurrent.*;

/**
 * This program is based on MergeSort.java in OpenJDK 7 example.
 */
public class MergeSort extends RecursiveAction {

    public static void main(String[] args) {
	int n = Integer.parseInt(args[0]);
	Random random = new Random(0);
	int[] array = new int[n];
	for (int i = 0; i < n; ++i) {
	    array[i] = random.nextInt();
	}
	long start = System.nanoTime();
	ForkJoinPool pool = new ForkJoinPool();
	pool.submit(new MergeSort(array, 0, array.length)).join();
	System.out.println(String.format("%f [msec]", (System.nanoTime() - start) / 1000000.0));
    }

    private final int[] array;
    private final int low;
    private final int high;

    MergeSort(int[] array, int low, int high) {
	this.array = array;
	this.low = low;
	this.high = high;
    }

    public void compute() {
	int size = high - low;
	if (size <= 8) {
	    Arrays.sort(array, low, high);
	}
	else {
	    int middle = low + (size >> 1);
	    invokeAll(new MergeSort(array, low, middle), new MergeSort(array, middle, high));
	    merge(middle);
	}
    }

    private void merge(int middle) {
	if (array[middle - 1] < array[middle]) {
	    return;
	}
	int copySize = high - low;
	int copyMiddle = middle - low;
	int[] copy = new int[copySize];
	System.arraycopy(array, low, copy, 0, copy.length);
	int p = 0;
	int q = copyMiddle;
	for (int i = low; i < high; ++i) {
	    if (q >= copySize || (p < copyMiddle && copy[p] < copy[q])) {
		array[i] = copy[p++];
	    }
	    else {
		array[i] = copy[q++];
	    }
	}
    }
}
